elephant/causality/granger.py

```diff
@@ -0,0 +1,438 @@
+
+
+def pairwise_granger(signals, max_order, information_criterion=aic):
+    r"""
+    Determine Granger Causality of two time series
+
+    Parameters
+    ----------
+    signals : np.ndarray or neo.AnalogSignal
+        time series data
+    max_order : int
+        maximal order of autoregressive model
+    information_criterion : callable
+        A function to compute the information criterion:
+            `bic` for Bayesian information_criterion,
+            `aic` for Akaike information criterion
+        Default: aic
+
+    Returns
+    -------
+    Causality
+        A `namedtuple` with the following attributes:
+            directional_causality_x_y : float
+                The Granger causality value for X influence onto Y.
+
+            directional_causality_y_x : float
+                The Granger causality value for Y influence onto X.
+
+            instantaneous_causality : float
+                The remaining channel interdependence not accounted for by
+                the directional causalities (e.g. shared input to X and Y).
+
+            total_interdependence : float
+                The sum of the former three metrics. It measures the dependence
+                of X and Y. If the total interdependence is positive, X and Y
+                are not independent.
+
+        Denote covariance matrix of signals
+            X by C|X  - a real number
+            Y by C|Y - a real number
+            (X,Y) by C|XY - a (2 \times 2) matrix
+
+        directional causality X -> Y given by
+            log(C|X / C|XY_00)
+        directional causality Y -> X given by
+            log(C|Y / C|XY_11)
+        instantaneous causality of X,Y given by
+            log(C|XY_00 / C|XY_11)
+        total interdependence of X,Y given by
+            log( {C|X \cdot C|Y} / det{C|XY} )
+
+    """
+    if isinstance(signals, AnalogSignal):
+        # transpose (N,2) -> (2, N)
+        signals = signals.magnitude.T
+    else:
+        signals = np.asarray(signals)
+
+    if not (signals.ndim == 2 and signals.shape[0] == 2):
+        raise ValueError("Required 2 x N input 'signals' array.")
+
+    # signal_x and signal_y are (1, N) arrays
+    signal_x, signal_y = np.expand_dims(signals, axis=1)
+
+    coeffs_x, var_x, p_1 = _optimal_vector_arm(signal_x, 1, max_order,
+                                               information_criterion)
+    coeffs_y, var_y, p_2 = _optimal_vector_arm(signal_y, 1, max_order,
+                                               information_criterion)
+    coeffs_xy, cov_xy, p_3 = _optimal_vector_arm(signals, 2, max_order,
+                                                 information_criterion)
+
+    sign, log_det_cov = np.linalg.slogdet(cov_xy)
+    tolerance = 1e-7
+
+    if sign <= 0:
+        raise ValueError(
+            "Determinant of covariance matrix must be always positive: "
+            "In this case its sign is {}".format(sign))
+
+    if log_det_cov <= tolerance:
```

Comment:
Is the consequence that the result may not be trusted?

Reply:
A comment was raised earlier that logdet has float precision issues. In that case we need to warn the user to keep in mind that the computation might not make sense. This approach is more suitable for batch processing.

Reply:
The idea, at least why I proposed to include such a tolerance, is not due to logdet having float precision issues (I am not aware of any; can the result of `np.logdet` be negative, say `-1e-8`? I don't know), but rather to warn the user when he (accidentally) passes a covariance matrix all values of which are close to zero. Is it better to warn the user immediately or ignore this rare case and proceed with the computation until an unexplainable error, if any, occurs at a later stage of processing?